src/Stratis.Bitcoin.Features.Consensus.Tests/Rules/ProvenHeaderRules/ProvenBlockHeaderCoinstakeRuleTest.cs

```diff
@@ -0,0 +1,532 @@
+            Action ruleValidation = () => this.consensusRules.RegisterRule<ProvenHeaderCoinstakeRule>().Run(this.ruleContext);
+            ruleValidation.Should().Throw<ConsensusErrorException>()
+                .And.ConsensusError
+                .Should().Be(ConsensusErrors.ReadTxPrevFailed);
+        }
+
+        [Fact]
+        public void RunRule_ProvenHeadersActive_And_CoinstakeUnspentOutputsIsNull_ReadTxPrevFailedErrorIsThrown()
+        {
+            // Setup proven header.
+            PosBlock posBlock = new PosBlockBuilder(this.network).Build();
+            ProvenBlockHeader provenBlockHeader = new ProvenBlockHeaderBuilder(posBlock, this.network).Build();
+            provenBlockHeader.Coinstake.Time = provenBlockHeader.Time;
+
+            // Setup chained header and move it to the height higher than proven header activation height.
+            this.ruleContext.ValidationContext.ChainedHeaderToValidate = new ChainedHeader(provenBlockHeader, provenBlockHeader.GetHash(), null);
+            this.ruleContext.ValidationContext.ChainedHeaderToValidate.SetPrivatePropertyValue("Height", this.options.ProvenHeadersActivationHeight + 10);
+
+            // Add more null unspent output to coinstake.
+            this.coinView
+                .Setup(m => m.FetchCoinsAsync(It.IsAny<uint256[]>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new FetchCoinsResponse(new[] { (UnspentOutputs)null }, posBlock.GetHash()));
+
+            // When we run the validation rule, we should hit coinstake read transaction error.
+            Action ruleValidation = () => this.consensusRules.RegisterRule<ProvenHeaderCoinstakeRule>().Run(this.ruleContext);
+            ruleValidation.Should().Throw<ConsensusErrorException>()
+                .And.ConsensusError
+                .Should().Be(ConsensusErrors.ReadTxPrevFailed);
+        }
+
+        [Fact]
+        public void RunRule_ProvenHeadersActive_And_CoinstakeIsIncorrectlySetup_NonCoinstakeErrorIsThrown()
+        {
+            // Setup proven header.
+            PosBlock posBlock = new PosBlockBuilder(this.network).Build();
+            ProvenBlockHeader provenBlockHeader = new ProvenBlockHeaderBuilder(posBlock, this.network).Build();
+
+            // Setup chained header and move it to the height higher than proven header activation height.
+            this.ruleContext.ValidationContext.ChainedHeaderToValidate = new ChainedHeader(provenBlockHeader, provenBlockHeader.GetHash(), null);
+            this.ruleContext.ValidationContext.ChainedHeaderToValidate.SetPrivatePropertyValue("Height", this.options.ProvenHeadersActivationHeight + 10);
+
+            // Setup coinstake transaction.
+            this.coinView
+                .Setup(m => m.FetchCoinsAsync(It.IsAny<uint256[]>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new FetchCoinsResponse(new[] { new UnspentOutputs(10, new Transaction()) }, posBlock.GetHash()));
```

Comment:
Does this need `.Verifiable()` ?  Or `Verify() or `VerifyAll()` at the end of the test ?  If you remove the setup the test is sucessfull.

Reply:
*If you remove the setup the test is still successful.

Reply:
I think its ok the setup is to make sure there is a return value.